lib/input-unmarshaller.js

```diff
@@ -27,46 +19,92 @@ module.exports = class InputUnmarshaller extends Transform {
 
     _transform(inputSignal, _, callback) {
         const dataSignal = inputSignal.data;
-        const inputIndex = dataSignal.argIndex || 0;
-        const rawPayload = dataSignal.payload;
+
+        const contentType = this._tryParseContentType(dataSignal, callback);
+        if (!contentType) {
+            return;
+        }
+        const unmarshalledPayload = this._tryUnmarshal(
+            contentType,
+            dataSignal,
+            callback
+        );
+        if (!unmarshalledPayload) {
+            return;
+        }
+
+        const finalPayload = this._tryConvertToMessage(
+            unmarshalledPayload,
+            dataSignal.headers,
+            callback
+        );
+        if (!finalPayload) {
+            return;
+        }
+
+        console.debug(`Forwarding data for input #${dataSignal.argIndex || 0}`);
+        try {
+            callback(null, finalPayload);
+        } catch (err) {
+            // propagate downstream error
+            this.emit("error", err);
+        }
+    }
+
+    _tryParseContentType(dataSignal, callback) {
         const contentType = dataSignal.contentType;
-        const acceptedContentType = determineContentTypes(contentType).accept;
-        if (!canUnmarshall(acceptedContentType)) {
+        try {
+            return parseContentType(contentType);
```

Comment:
a method should be either sync or async, never both.

Reply:
I could inline these extracted methods again, but then the flow becomes quite big and less readable which is why I extracted these functions in the first place

Reply:
Stuck with synchronous style. 
I changed the code to keep the extracted methods but they now return either a value or a RiffError.
This at least remains 100% synchronous. 
I could go further and introduce a Either monad or something, but I'm not sure it's worth the effort.